src/combinatorics/quiver/symmetric.lean

```diff
@@ -0,0 +1,212 @@
+  p.reverse.reverse = p :=
+begin
+  induction p,
+  { simp, },
+  { simp only [path.reverse, path.reverse_comp, path.reverse_to_path, reverse_reverse, p_ih],
+    refl, },
+end
+
+namespace symmetrify
+
+/-- The inclusion of a quiver in its symmetrification -/
+def of : prefunctor V (symmetrify V) :=
+{ obj := id,
+  map := λ X Y f, sum.inl f }
+
+/-- Given a quiver `V'` with reversible arrows, a prefunctor to `V'` can be lifted to one from
+    `symmetrify V` to `V'` -/
+def lift {V' : Type*} [quiver V'] [has_reverse V'] (φ : V ⥤q V') :
+  (symmetrify V) ⥤q V' :=
+{ obj := φ.obj,
+  map := λ X Y f, sum.rec (λ fwd, φ.map fwd) (λ bwd, reverse (φ.map bwd)) f }
+
+lemma lift_spec  (V' : Type*) [quiver V'] [has_reverse V'] (φ : V ⥤q V') :
+  of ⋙q (lift φ) = φ :=
+begin
+  fapply prefunctor.ext,
+  { rintro X, refl, },
+  { rintros X Y f, refl, },
+end
+
+lemma lift_reverse  (V' : Type*) [quiver V'] [h : has_involutive_reverse V']
+  (φ : V ⥤q V')
+  {X Y : symmetrify V} (f : X ⟶ Y) :
+  (lift φ).map (quiver.reverse f) = quiver.reverse ((lift φ).map f) :=
+begin
+  dsimp [lift], cases f,
+  { simp only, refl, },
+  { simp only, rw h.inv', refl, }
+end
+
+/-- `lift φ` is the only prefunctor extending `φ` and preserving reverses. -/
+lemma lift_unique (V' : Type*) [quiver V'] [has_reverse V']
+  (φ : V ⥤q V')
+  (Φ : (symmetrify V) ⥤q V')
+  (hΦ : of ⋙q Φ = φ) [hΦrev : Φ.map_reverse] :
+  Φ = lift φ :=
+begin
+  subst_vars,
+  fapply prefunctor.ext,
+  { rintro X, refl, },
+  { rintros X Y f,
+    cases f,
+    { refl, },
+    { dsimp [lift,of],
+      simp only [←prefunctor.map_reverse', symmetrify_reverse, sum.swap_inl], }, },
+end
+
+/-- A prefunctor canonically defines a prefunctor of the symmetrifications. -/
+@[simps] def _root_.prefunctor.symmetrify (φ : U ⥤q V) : (symmetrify U) ⥤q (symmetrify V) :=
+{ obj := φ.obj,
+  map := λ X Y, sum.map φ.map φ.map }
+
+instance _root_.prefunctor.symmetrify_map_reverse  (φ : U ⥤q V) :
+  prefunctor.map_reverse φ.symmetrify := ⟨λ u v e, by { cases e; refl }⟩
+
+end symmetrify
+
+namespace push
+
+variables {W} (σ : V → W)
+
+instance [has_reverse V] : has_reverse (push σ) :=
+{ reverse' := λ a b F, by { cases F, constructor, apply reverse, exact F_f, } }
+
+instance [h : has_involutive_reverse V] : has_involutive_reverse (push σ) :=
+{ reverse' := λ a b F, by { cases F, constructor, apply reverse, exact F_f, },
+  inv' :=  λ a b F, by { cases F, dsimp [reverse], congr, apply h.inv', } }
+
+lemma of_reverse [h : has_involutive_reverse V]  (X Y : V) (f : X ⟶ Y):
+  (reverse $ ((push.of σ)).map f) = ((push.of σ)).map (reverse f) := rfl
+
+instance of_map_reverse [h : has_involutive_reverse V] : (push.of σ).map_reverse :=
+⟨ by simp [of_reverse] ⟩
+
+end push
+
+/--
+A quiver with reverses is preconnected iff there exists a path between any pair of
+vertices.
+-/
+@[nolint unused_arguments]
+def is_preconnected (V) [quiver.{u+1} V] [has_reverse V] := ∀ (X Y : V), nonempty (path X Y)
```

Comment:
Why assuming `has_reverse`? Without it, you can talk about directed-connected quivers.

Reply:
My thinking was that the meaning is less clear when the quiver doesn't have reverses, so that putting this as an assumption forces one to use it only when the meaning is clear. I can remove it.

Reply:
The rational was to put everything related to `has_reverse` (but not deserving its own file) in this file essentially. Where else would you see it?

Reply:
Notice that [category_theory.is_preconnected](https://leanprover-community.github.io/mathlib_docs/category_theory/is_connected.html#category_theory.is_preconnected) does not require morphisms both way.